products/libllbuild/BuildDB-C-API.cpp

```diff
@@ -0,0 +1,168 @@
+//===-- BuildDB-C-API.cpp ---------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2015 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include <llbuild/llbuild.h>
+
+#include "llbuild/Core/BuildDB.h"
+
+#include <mutex>
+
+using namespace llbuild;
+using namespace llbuild::core;
+
+namespace {
+  
+  class CAPIBuildDBKeysResult {
+  private:
+    std::vector<KeyType> keys;
+    
+  public:
+    CAPIBuildDBKeysResult(std::vector<KeyType> keys): keys(keys) { }
+    
+    CAPIBuildDBKeysResult(const CAPIBuildDBKeysResult&) LLBUILD_DELETED_FUNCTION;
+    CAPIBuildDBKeysResult& operator=(const CAPIBuildDBKeysResult&) LLBUILD_DELETED_FUNCTION;
+    
+    const uint32_t size() {
+      return keys.size();
+    }
+    
+    const KeyType keyAtIndex(KeyID index) {
+      return keys[index];
+    }
+  };
+  
+  class CAPIBuildDB: public BuildDBDelegate {
+    /// The key table, for memory caching of key to key id mapping.
+    llvm::StringMap<KeyID> keyTable;
+    
+    /// The mutex that protects the key table.
+    std::mutex keyTableMutex;
+    
+    std::unique_ptr<BuildDB> _db;
+    
+  public:
+    CAPIBuildDB(StringRef path, uint32_t clientSchemaVersion, std::string *error_out) {
+      _db = createNondestructiveSQLiteBuildDB(path, clientSchemaVersion, error_out);
+      if (error_out != NULL) {
+        return;
+      }
+      // The delegate  caches keys for key ids
+      _db.get()->attachDelegate(this);
+      
+      if (!buildStarted(error_out)) {
+        buildComplete();
+        return;
+      }
+    }
+    
+    virtual const KeyID getKeyID(const KeyType& key) override {
```

Comment:
Let's have a reference here to the BuildEngine copy of the same code, for consistency if one of them changes.